src/clients/UBAClient/UBAClientWithManual.ts

```diff
@@ -0,0 +1,119 @@
+// THIS CLASS HAS NO PURPOSE RIGHT NOW - IT WILL BE UPDATED IN THE FUTURE
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import assert from "assert";
+import winston from "winston";
+import { RefundRequestWithBlock, UbaFlow } from "../../interfaces";
+import { isDefined } from "../../utils";
+import { BaseUBAClient, OpeningBalanceReturnType, RequestValidReturnType } from "./UBAClientAbstract";
+import { BigNumber } from "ethers";
+import { RelayerFeeDetails } from "../../relayFeeCalculator";
+
+/**
+ * UBAClientState is a type that represents the state of a UBAClient.
+ */
+type UBAClientState = {
+  spoke: {
+    [chainId: number]: {
+      flows: UbaFlow[];
+      latestBlockNumber: number;
+      bundleEndBlocks: number[];
+      openingBalances: {
+        [token: string]: {
+          blockNumber: number;
+          balance: BigNumber;
+        }[];
+      };
+    };
+  };
+};
+
+/**
+ * UBAClient is a base class for UBA functionality. It provides a common interface for UBA functionality to be implemented on top of or extended.
+ * It is intended to only be used directly when RPC calls are __***not***__ required, e.g. in tests, Scraper API.
+ */
+export class UBAClientManual extends BaseUBAClient {
+  /**
+   * Constructs a UBAClient instance.
+   * @param chainIdIndices A list of available chain indices that should be reasoned about
+   * @param state Manual state that acts as the source of truth in this model. Note that these are initial values and can be modified by extended class features
+   * @param logger Optional logger to be used for logging
+   */
+  constructor(
+    readonly chainIdIndices: number[],
+    protected readonly state: UBAClientState,
+    readonly logger?: winston.Logger
+  ) {
+    super(chainIdIndices, logger);
+    assert(chainIdIndices.length > 0, "No chainIds provided");
+    assert(Object.values(state.spoke).length > 0, "No SpokePools provided");
+  }
+
+  protected resolveClosingBlockNumber(chainId: number, blockNumber: number): number {
+    return this.state.spoke[chainId].bundleEndBlocks.find((bundleEndBlock) => bundleEndBlock <= blockNumber) ?? -1;
```

Comment:
Not sure how you wanna reconcile this code with the very similar function from `relayer-v2` [here](https://github.com/across-protocol/relayer-v2/blob/master/src/dataworker/DataworkerUtils.ts#L51)

I think your code is a lot cleaner FWIW.

Reply:
I think we can leave this as is because the relayer-v2 code has logging baked into it